src/Storages/MergeTree/MergeTreeReaderTextIndex.cpp

```diff
@@ -140,6 +140,17 @@ bool MergeTreeReaderTextIndex::canSkipMark(size_t mark, size_t current_task_last
         analyzed_granules.add(index_mark);
     }
 
+    if (const auto * data_part = typeid_cast<const LoadedMergeTreeDataPartInfoForReader *>(data_part_info_for_read.get());
+        data_part->getDataPart()->storage.merging_params.mode == MergeTreeData::MergingParams::Replacing)
```

Comment:
a follow-up would be passing the `is_final` boolean parameter to check this condition only when `is_final` is set to true. It requires a bit more change, therefore it can be postponed to a follow-up PR.